test/controllers/registrations_controller_test.rb

```diff
@@ -55,12 +55,55 @@ class RegistrationsControllerTest < ActionDispatch::IntegrationTest
       ]
     )
 
-    post(
-      callback_registration_url,
-      params: { credential_nickname: "USB Key" }.merge(public_key_credential)
-    )
+    assert_no_difference -> { User.count } do
+      post(
+        callback_registration_url,
+        params: { credential_nickname: "USB Key" }.merge(public_key_credential)
+      )
+    end
 
     assert_response :unprocessable_entity
     assert_equal "Couldn't register your Security Key", response.body
   end
+
+  test "should register successfully" do
+    authenticator = WebAuthn::FakeAuthenticator.new
+
+    fake_rp = WebAuthn::RelyingParty.new(
+      origin: "https://fake.relying_party.test",
+      id: "fake.relying_party.test",
+      name: "Fake RelyingParty"
+    )
+
+    fake_client = WebAuthn::FakeClient.new("https://fake.relying_party.test", authenticator:)
+
+    user = User.new(username: "John Doe")
+
+    raw_challenge = SecureRandom.random_bytes(32)
+    challenge = WebAuthn.configuration.encoder.encode(raw_challenge)
+
+    webauthn_credential = fake_client.create(challenge:, rp_id: fake_rp.id, user_verified: true)
+
+    session_data = {
+      current_registration: {
+        challenge:,
+        user_attributes: user.attributes
+      }
+    }
+
+    ApplicationController.stub_any_instance(:relying_party, fake_rp) do
+      RegistrationsController.stub_any_instance(:session, session_data) do
+        assert_difference 'User.count', +1 do
+          assert_difference 'Credential.count', +1 do
+            post(
+              callback_registration_url,
+              params: { credential_nickname: "USB Key" }.merge(webauthn_credential)
+            )
+          end
+        end
+      end
+    end
+
+    assert_response :success
+  end
```

Comment:
What do you think about using a similar approach to the one used in the already existent tests?

```ruby
  test "should register successfully" do
    raw_challenge = SecureRandom.random_bytes(32)
    challenge = WebAuthn.configuration.encoder.encode(raw_challenge)

    WebAuthn::PublicKeyCredential::CreationOptions.stub_any_instance(:raw_challenge, raw_challenge) do
      post registration_url, params: { registration: { username: "alice" }, format: :json }

      assert_response :success
    end

    public_key_credential =
      WebAuthn::FakeClient
      .new(Rails.configuration.webauthn_origin)
      .create(challenge:, user_verified: true)

    webauthn_credential = WebAuthn::Credential.from_create(public_key_credential)

    assert_difference 'User.count', +1 do
      assert_difference 'Credential.count', +1 do
        post(
          callback_registration_url,
          params: { credential_nickname: "USB Key" }.merge(public_key_credential)
        )
      end
    end

    assert_response :success
  end
```

Reply:
Sure, we can re-use your way for consistency's sake.

That PR is an extract from my rewrite of your demo. I'm trying to understand how Passkey is implemented.

In my test, I mocked the `relying_party` itself so I could call the controller as normal. 

In your test, you mock `WebAuthn::PublicKeyCredential::CreationOptions`,  which is unclear to me how it makes `verify_registration` pass as that is an internal component of `WebAuthn` that the test should not be aware of.

```ruby
describe "POST #callback" do

  context "with valid params" do

    let(:fake_rp) do
      WebAuthn::RelyingParty.new(
        origin: "https://fake.relying_party.test",
        id: "fake.relying_party.test",
        name: "Fake RelyingParty"
      )
    end

    let(:fake_client) do
      WebAuthn::FakeClient.new("https://fake.relying_party.test", authenticator:)
    end

    it "registers successfully" do
      fake_challenge = WebAuthn.configuration.encoder.encode(SecureRandom.random_bytes(32))

      webauthn_credential = fake_client.create(challenge: fake_challenge, rp_id: fake_rp.id, user_verified: true)

      allow(controller).to receive(:relying_party).and_return(fake_rp)

      post(:callback,
              params: {
                credential_nickname: "John's Security Key",
                user: user.to_h.slice(:id, :name),
                authenticator_selection: { user_verification: "required" }
              }.merge(webauthn_credential),
              session:,
              format: :json
            )

      expect(response).to be_successful
    end
  end
end
```

Reply:
Hey!

Sure, I can explain 🙂 

In essence, both our approaches are just different approaches for mocking the challenge that's essential for the verifying process of WebAuthn to succeed. The challenge is a critical part of the registration and authentication process, as it used for preventing replying attacks.

In the tests that we have on the demo, we are mocking `WebAuthn::PublicKeyCredential::CreationOptions` because we are hitting the `create` action of the `RegistrationController` [which calls `RelyingParty#options_for_registration`](https://github.com/cedarcode/webauthn-rails-demo-app/blob/6f0c143/app/controllers/registrations_controller.rb#L10), asks for the challenge to the returned object and stores that challenge on the session. That 'returned object' is simply a `WebAuthn::PublicKeyCredential::CreationOptions` (`RelyingParty#options_for_registration` [calls `WebAuthn::Credential.options_for_create`](https://github.com/cedarcode/webauthn-ruby/blob/b90c6fd/lib/webauthn/relying_party.rb#L85) which [returns a new `WebAuthn::PublicKeyCredential::CreationOptions` with the params received](https://github.com/cedarcode/webauthn-ruby/blob/5f84dfc/lib/webauthn/credential.rb#L11)) and that is why we mock any instance of it to receive `challenge` and return a fixed challenge that we have created before.

The crucial point to understand here is that by mocking `WebAuthn::PublicKeyCredential::CreationOptions`, we directly influence the challenge that's going to be stored in the session. This is essential as we then use the fake client to create a credential using that same challenge and send those params to the `callback` action, ensuring that both the credential that will come in the request params and the session will have the same challenge.

On the other hand, in your approach, you achieve the same result by simply [mocking the session](https://github.com/cedarcode/webauthn-rails-demo-app/blob/1afd495b0857bb69b167822a98bc4d2c2d5e47f9/test/controllers/registrations_controller_test.rb#L95) to have the same fixed challenge that you use for creating the credential (using the fake client) so that you ensure that both the credential that will come in the request params and the session will have the same challenge.

Worth noting that the reason that you mock the relying party is because you need the relying party's `origin` to match your fake client's `origin`. We are currently doing that by simply instantiating the fake client with our environment's origin which is the same that [will be used for instantiating the relying party in the controller](https://github.com/cedarcode/webauthn-rails-demo-app/blob/b2cbeae/test/controllers/registrations_controller_test.rb#L41).

Having said that, the important thing to note is that both approaches work so feel free to use the one that works best for you 🙂 

I hope I explained myself! Feel free to reach out if you have any more questions or need further clarification.